### DISCUSSION TITLE
Error with infer when creating zod wrapper (TS2503: Cannot find namespace)

### DISCUSSION THREAD

**Discussion by disillusionedowl** [colinhacks/zod#3005] [Q&A] [answered]
Hi, am I doing something wrong please? The end goal is to wrap zod so I can swap/adjust validator implementations, decorate with custom logging and other functions or methods, etc. I am new to JS/TS. Thanks in advance!

```js
import { z } from 'zod';

const myZ = z;

const SomeSchema = myZ.strictObject( {
	'type': myZ.union( [
		myZ.literal( 'hello' ),
		myZ.literal( 'world' )
	] )
} ); // <=== works

console.log( SomeSchema ); // <=== works

export type SomeType = myZ.infer<typeof SomeSchema>;  // <=== ERROR TS2503: Cannot find namespace 'myZ'
```

Here is my tsconfig.json:

```json
{
    "include": ["**/*.ts", "**/*.tsx", "types/"],
    "compilerOptions": {
        "target": "ES6",
        "moduleResolution": "node",
        "strict": true,
        "jsx": "react-jsx",
        "esModuleInterop": true,
        "plugins": [{ "name": "typescript-plugin-css-modules" }]
    }
}
```

**Comment by JacobWeisenburger** [comment]
overriding parts of zod with your own implementation is not recommended and will lead to lots of errors.

**Comment by disillusionedowl** [comment]
@JacobWeisenburger Thank you, that is probably true although it would be great if at least a simple wrapper around the `z` library would work. Let's say a wrapper that does whatever `z` does but also `console.log` these calls or something at the same time (just an example). I am used to this design pattern from the more classic OOP languages. I like the idea of having such centralized connection between my and 3rd party code. But like I said, many JS/TS intricacies still escape me, so maybe I am just missing something and it's not hard/impossible to do it?

**Comment by JacobWeisenburger** [comment]
What would you like to have happen with this issue? are you wanting it to stay open so it may eventually be added to Zod? or are you looking for a work around?

**Comment by disillusionedowl** [comment]
I think I am way in the minority here so unless it's objectively really interesting it can be closed as far as I am concerned. Although not what I envisioned initially, I can always just interface with Zod directly. Workaround would be nice, just so I can learn more about JS/TS but obviously don't want to take too much of anyone's time on such a low issue. Thank you either way @JacobWeisenburger

**Reply by JacobWeisenburger** [reply]
I'm happy to help with work arounds or questions. But your question seems very abstract, would you be able to make it more specific? then I might be able to help you more.

**Reply by disillusionedowl** [reply]
Thanks, Jacob. Here is a working example of how I am wrapping `useReducer` from React instead of using it directly.

```tsx
// FILE: useMyReducer.tsx
import { useEffect, useReducer } from 'react';

import {
	MyReducer,
	MyStateInitialProps,
	MyStateType
} from '../../';

export const useMyReducer = () => {
	const [ myState, myDispatch ] = useReducer( MyReducer, MyStateInitialProps )

	const myDispatchWrapper: typeof myDispatch = ( ...props ) => { // the wrapper function
		console.log( 'oldState:', myState );
		console.log( 'reducerAction:', ...props );
		return myDispatch( ...props ); // the original function
	}
	
	useEffect( () => {
		console.log( 'newState:', myState );
	}, [ myState ] );

	return { myState, myDispatch: myDispatchWrapper }; // repackaged to have exactly the same output as the original useReducer()
}
```

The above is a common design pattern and allows me to e.g. `console.log` the new/old state anytime it changes via a reducer (**from anywhere**) and various other tweaks I could think of later. `useReducer` is a function so it is easy to wrap it and pass through the props. In OOP with class objects, I would do that by extending the class and defining a common interface. And with `z` it is apparently not so easy or I don't know the correct way in JS/TS. Does it make better sense now what I mean when I say "wrapper"?

**Reply by JacobWeisenburger** [reply]
Sorry, I wasn't clear. I do understand what a wrapper is, however I'm not sure how you are wanting to use it with Zod. Please send an example of how you would do this with Zod. Pseudo code is fine, if that's easier

**Reply by disillusionedowl** [reply]
I didn't have any immediate use for it, but doing some pseudo OOP code usecase in different architecture as an example:

`Zod` can change API at some point in the future (e.g. rename some random method `delete` to `remove`) I could then add these 3 middle lines and have it instantly hotfixed, assuming I have coded my codebase towards the common ValidatorInterface and provided instances of MyValidator instead of ZodValidator to the consumers:

```
class MyValidator extends ZodValidator implements ValidatorInterface {
	function delete( ...props ) {
		return parent->remove( ...props ) );
	}
        // ... all the other methods are automatically inherited and passed through out-of-the-box, unless overridden as the one above
}
```

or I could even use a different validator altogether if I manually rewire all the methods/calls since the codebase is coded towards a common interface and not concrete implementation. I am not saying these things are a must (or even clean) but it's nice to have options like this if needed and have the codebase somewhat decoupled.

**Reply by JacobWeisenburger** [reply]
I can't really see how your example would decouple your codebase from zod. this seems like it would add a ton of overhead to override every method provided by zod with your own implementation.

**Reply by JacobWeisenburger** [reply]
personally, I'm not a huge fan of OOP, I would highly recommend taking more of a functional approach and just making pure functions for whatever parts of zod you need to wrap.

**Reply by disillusionedowl** [reply]
The idea is to always wrap the whole thing from the start and make tweaks as the project progresses if needed (or not). That is pretty easy to do in OOP because you just pass it through as-is at the beginning (you don't really need to type out the whole API, although you could) so there is very little overhead at that point in time.

Another usecase is you wrap a logger lib, have it log some variables here and there but then you decide you want all these logs to also contain information about the filepath (backtrace). If you setup your project as I am suggesting from the start, you don't need to go change every call to that lib but instead you can simply override the parent method like this from one central place:

```php
class MyLogger extends MonologLogger implements LoggerInterface {
	function addRecord( ...$args ): bool {
		return parent::addRecord( ...\array_replace( $args, [
			1 => $args[1].' | '.debug_backtrace( DEBUG_BACKTRACE_IGNORE_ARGS, 4 )[3][ 'class' ]
		] ) );
	}
}
```

Personally I enjoy simple OOP design patterns that are also applicable in a more functional style as you have suggested (it's just a thought pattern after all) and dislike classes (class inheritance and props mutation are archaic patterns imo). The problem is I wasn't able to do this with Zod unfortunately because it got stuck on `.infer` (TS2503: Cannot find namespace) for some reason and I can't tell if it's really a Zod issue or even general JS/TS issue. What exactly is the imported `z` anyway, is it an object/module/namespace/type/combination? Apologies for such basic questions, many JS/TS concepts are still new to me.

**Reply by JacobWeisenburger** [reply]
Seems fairly easy to do something like this for what you are asking:

```ts
function myZodParse<Schema extends z.Schema> ( schema: Schema, data: z.infer<Schema> ) {
    // do something else
    return schema.parse( data )
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by disillusionedowl** [reply, marked as answer]
Thanks for that @JacobWeisenburger , although I would like to wrap the whole thing and not just one function.

I think I have a somewhat working solution now (see below). Unless you see any issues with this approach, I think I am ready to mark it as solved.

`validator.tsx :`
```tsx
import { z } from 'zod';

export * from 'zod';
export type sayHiType = () => void;
export const sayHi:sayHiType = () => console.log( 'Hi!' );
export const strictObject: typeof z.strictObject = ( ...props ) => {
	console.log( 'Hello with strictObject!' );
	return z.strictObject( ...props );
}

```

`index.ts :`
```tsx
export * as validator from './validator';
```

`consumer.tsx :`
```tsx
import { validator } from './';

const SomeSchema = validator.strictObject( {
	'type': validator.union( [
		validator.literal( 'hello' ),
		validator.literal( 'world' )
	] )
} ); // => works, logs that 'Hello with strictObject!' string as well

console.log( SomeSchema ); // => works
validator.sayHi(); // => works
export type SomeType = validator.infer<typeof SomeSchema>; // => works
```

-------------------------
Some extra thoughts:

As you can see, it's just a few lines of code to implement such wrapper, so there is not so much complexity from the start. Consumer is basically unaware of the validator implementation, could be zod, yup, typebox, etc behind the scenes as long the API matches. The API can further be rewired/extended in `validator.tsx` if needed in the future.

It would also be great to get rid of this

```tsx
export * as validator from './validator';
```
line from `index.ts` file, in favor of the more classic line:

```tsx
export * from './validator';
```

and name the "validator" elsewhere, ideally inside of that `validator.tsx`, but I don't think that is possible.

**Reply by JacobWeisenburger** [reply]
it seems like you have answered your own question. is that correct?

**Reply by disillusionedowl** [reply]
Looks good enough to me so unless there are any objections to this approach, I will consider it as solved. Thanks for your help!

**Reply by JacobWeisenburger** [reply]
You're welcome.

Please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
